extmod/modbluetooth.c

```diff
@@ -1161,6 +1162,72 @@ STATIC mp_obj_t invoke_irq_handler(uint16_t event,
     return result;
 }
 
+#if MICROPY_PY_THREAD
+
+// If threading is enabled with BLE synchronous events then the BLE event callbacks
+// may occur on any system thread (including a MicroPython thread), and must be
+// synchronised with the rest of the MicroPython runtime.
+
+#if MICROPY_ENABLE_PYSTACK
+#error not supported
+#endif
+
+STATIC mp_obj_t invoke_irq_handler(uint16_t event,
+    const mp_int_t *numeric, size_t n_unsigned, size_t n_signed,
+    const uint8_t *addr,
+    const mp_obj_bluetooth_uuid_t *uuid,
+    const uint8_t **data, size_t *data_len, size_t n_data) {
+
+    // This code may run on an existing MicroPython thread, or a non-MicroPython thread
+    // that's not using the mp_thread_get_state() value.  In the former case the state
+    // must be restored once this callback finishes.
+    mp_state_thread_t *ts_orig = mp_thread_get_state();
+
+    mp_state_thread_t ts;
+    if (ts_orig == NULL) {
+        mp_thread_set_state(&ts);
+        mp_stack_set_top(&ts + 1); // need to include ts in root-pointer scan
+        mp_stack_set_limit(MICROPY_PY_BLUETOOTH_IRQ_STACK_SIZE - 1024);
+        ts.gc_lock_depth = 0;
+        ts.mp_pending_exception = MP_OBJ_NULL;
+        mp_locals_set(mp_state_ctx.thread.dict_locals); // set from the outer context
+        mp_globals_set(mp_state_ctx.thread.dict_globals); // set from the outer context
+        MP_THREAD_GIL_ENTER();
+    }
+
+    mp_obj_t result = mp_const_none;
+    nlr_buf_t nlr;
+    if (nlr_push(&nlr) == 0) {
+        result = invoke_irq_handler_run(event, numeric, n_unsigned, n_signed, addr, uuid, data, data_len, n_data);
```

Comment:
might need to add `mp_sched_lock()` and `mp_sched_unlock()` around this call, to make sure the VM does not go off and handle other tasks

Reply:
done